AAInstrumentation/optimization.py

```diff
@@ -180,6 +180,71 @@ def autotuner_factory():
     return lambda driver: AutoTuningOptimizer(driver)
 
 
+class ParallelAutotuner(Optimizer):
+    """
+    This class implements the local Autotuner in a parallel way. It aims to
+    optimize the size of the final binary by sequentially trying out the queries
+    in the compilation. If a query reduces the size, it is fied to imprecise for
+    all following queries.
+
+    This implementation is parallelized in that it proceeds batch-wise. For a
+    given batch, each individual query is relaxed in parallel. If no query
+    reduces the size, the next batch is tried. If query Q reduces the size, the
+    next batch is started with Q relaxed and from the index of Q. If multiple
+    queries reduce the size, the first one is treated as Q in the previous case.
+
+    """
+
+    def __init__(self, driver: OptimizerDriver):
+        super().__init__(driver)
+        self.batch_size = driver.proc_count
+
+    description = "Parallel AutoTuner"
+    batch_size: int = 0
+
+    def handle_batch_from_index(
+        self, file_name: Path, prefix: list[int], index: int
+    ) -> list[int]:
+        sizes = []
+        with Pool(self.batch_size) as pool:
+            sizes = pool.starmap(
+                self.driver.run_assemble_and_measure_file,
+                [
+                    (file_name, i, prefix + [i])
+                    for i in range(index, index + self.batch_size)
+                ],
+            )
+
+        return sizes
+
+    def update_iteration(
+        self, file_name: Path, prefix: list[int], index: int, current_minimum: int
+    ) -> tuple[list[int], int, int]:
+        sizes = self.handle_batch_from_index(file_name, prefix, index)
+
+        for i, size in enumerate(sizes):
+            if size < current_minimum:
+                return prefix + [index + i], index + i, size
```

Comment:
I'm not sure if it makes sense, but I think the code would be much more readable if there were explicit types for these AA sequences, then also the return types won't be as cryptic as `tuple[list[int], int, int]`

Reply:
I decided to only this for this specific optimizer atm. I will file a follow up PR applying this to the rest, once this is merged.